### ISSUE TITLE
Calico-node couldn't set routes when network interface is restarted

### ISSUE THREAD

**Issue by mdianjun** [projectcalico/calico#1658]
My operations:
1. on a node, execute ifdown bond0; ifup bond0
2. create a pod on the node

The routes  during the above operations:
```
[root@KC-BCC-POD22-10-196-102-130 ~]# ip route
default via 10.196.102.129 dev bond0 
blackhole 10.163.32.0/28  proto bird 
10.163.32.4 dev calif61679f6dfe  scope link    //the pod's route appeard
10.163.32.15 dev cali867f5376d81  scope link 
10.196.102.128/27 dev bond0  proto kernel  scope link  src 10.196.102.130 
169.254.0.0/16 dev bond0  scope link  metric 1004 
192.168.0.0/16 dev docker0  proto kernel  scope link  src 192.168.0.1 
[root@KC-BCC-POD22-10-196-102-130 ~]# ip route //the pod's route disappeard
default via 10.196.102.129 dev bond0 
blackhole 10.163.32.0/28  proto bird 
10.163.32.15 dev cali867f5376d81  scope link 
10.196.102.128/27 dev bond0  proto kernel  scope link  src 10.196.102.130 
169.254.0.0/16 dev bond0  scope link  metric 1004 
192.168.0.0/16 dev docker0  proto kernel  scope link  src 192.168.0.1 
```

The calico-node log during the above operations:
```
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 447: Linux interface addrs changed. addrs=set.mapSet{} ifaceName="calif61679f6dfe"
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 598: Received interface addresses update msg=&intdataplane.ifaceAddrsUpdate{Name:"calif61679f6dfe", Addrs:set.mapSet{}}
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 690: Applying dataplane updates
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 704: Finished applying updates to dataplane. msecToApply=0.063861
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 447: Linux interface addrs changed. addrs=set.mapSet{"fe80::ecee:eeff:feee:eeee":set.empty{}} ifaceName="calif61679f6dfe"
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 429: Linux interface state changed. ifaceName="calif61679f6dfe" state="up"
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 598: Received interface addresses update msg=&intdataplane.ifaceAddrsUpdate{Name:"calif61679f6dfe", Addrs:set.mapSet{"fe80::ecee:eeff:feee:eeee":set.empty{}}}
2018-02-02 09:13:19.733 [INFO][153] int_dataplane.go 690: Applying dataplane updates
2018-02-02 09:13:19.734 [INFO][153] int_dataplane.go 704: Finished applying updates to dataplane. msecToApply=0.07749199999999999
2018-02-02 09:13:19.734 [INFO][153] int_dataplane.go 588: Received interface update msg=&intdataplane.ifaceUpdate{Name:"calif61679f6dfe", State:"up"}
2018-02-02 09:13:19.734 [INFO][153] int_dataplane.go 690: Applying dataplane updates
2018-02-02 09:13:19.734 [INFO][153] endpoint_mgr.go 239: Workload interface came up, marking for reconfiguration. ifaceName="calif61679f6dfe"
2018-02-02 09:13:19.734 [INFO][153] endpoint_mgr.go 777: Applying /proc/sys configuration to interface. ifaceName="calif61679f6dfe"
2018-02-02 09:13:19.735 [INFO][153] route_table.go 442: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.163.32.4/32 ifaceName="calif61679f6dfe" ipVersion=0x4
2018-02-02 09:13:19.735 [INFO][153] route_table.go 298: Interface in cleanup grace period, will retry after. ifaceName="calif61679f6dfe" ipVersion=0x4
2018-02-02 09:13:19.735 [INFO][153] int_dataplane.go 704: Finished applying updates to dataplane. msecToApply=0.9092049999999999
2018-02-02 09:13:20.900 [INFO][153] iface_monitor.go 168: Netlink address update. addr="fe80::ecee:eeff:feee:eeee" exists=true ifIndex=168
2018-02-02 09:13:20.911 [INFO][153] health.go 150: Overall health summary=&health.HealthReport{Live:true, Ready:true}
2018-02-02 09:13:21.077 [INFO][153] health.go 150: Overall health summary=&health.HealthReport{Live:true, Ready:true}
2018-02-02 09:13:21.424 [INFO][153] int_dataplane.go 690: Applying dataplane updates
2018-02-02 09:13:21.424 [INFO][153] ipsets.go 222: Asked to resync with the dataplane on next update. family="inet"
2018-02-02 09:13:21.425 [INFO][153] ipsets.go 253: Resyncing ipsets with dataplane. family="inet"
2018-02-02 09:13:21.425 [INFO][153] route_table.go 442: Syncing routes: found unexpected route; ignoring due to grace period. dest=10.163.32.4/32 ifaceName="calif61679f6dfe" ipVersion=0x4
2018-02-02 09:13:21.425 [INFO][153] route_table.go 298: Interface in cleanup grace period, will retry after. ifaceName="calif61679f6dfe" ipVersion=0x4
2018-02-02 09:13:21.426 [INFO][153] ipsets.go 295: Finished resync family="inet" numInconsistenciesFound=0 resyncDuration=1.08227ms
2018-02-02 09:13:21.426 [INFO][153] int_dataplane.go 704: Finished applying updates to dataplane. msecToApply=1.3420649999999998

2018-02-02 09:13:30.911 [INFO][153] health.go 150: Overall health summary=&health.HealthReport{Live:true, Ready:true}
2018-02-02 09:13:31.077 [INFO][153] health.go 150: Overall health summary=&health.HealthReport{Live:true, Ready:true}
2018-02-02 09:13:31.719 [INFO][153] int_dataplane.go 690: Applying dataplane updates
2018-02-02 09:13:31.719 [INFO][153] ipsets.go 222: Asked to resync with the dataplane on next update. family="inet"
2018-02-02 09:13:31.719 [INFO][153] ipsets.go 253: Resyncing ipsets with dataplane. family="inet"
2018-02-02 09:13:31.719 [INFO][153] route_table.go 446: Syncing routes: removing old route. dest=10.163.32.4/32 ifaceName="calif61679f6dfe" ipVersion=0x4   //the route is removed!!!!
2018-02-02 09:13:31.719 [INFO][153] conntrack.go 78: Removing conntrack flows ip=10.163.32.4
2018-02-02 09:13:31.720 [INFO][153] ipsets.go 295: Finished resync family="inet" numInconsistenciesFound=0 resyncDuration=1.073688ms
2018-02-02 09:13:31.720 [INFO][153] int_dataplane.go 704: Finished applying updates to dataplane. msecToApply=1.444914
```
Pay attention the below log, route is removed:
```
Syncing routes: removing old route. dest=10.163.32.4/32 ifaceName="calif61679f6dfe" ipVersion=0x4
```

Why calico-node removed route when network interface was restartd?

calico version: 3.0.1
kubernetes version: 1.8.4

**Comment by fasaxc** [comment]
It looks like the CNI plugin added the route but Felix didn't hear about the new endpoint.  If that happens, Felix waits for a while and then removes the route.  There are a couple of possibilities:

- Felix lost contact with the datastore but didn't spot it, if you're using etcdv3, we're investigating an issue that could cause that under https://github.com/projectcalico/calico/issues/1632; the workaround is to restart calico/node if that happens
- The CNI plugin failed to write the endpoint to the datastore, which would give an error in the kubelet log.

**Comment by mdianjun** [comment]
@fasaxc   I tried the experiment below:
Don't use calico-etcd clusterIP, but use  node ip for calico-node.  Configure it in calico-config, like this `etcd_endpoints: "http://10.196.104.195:6666,http://10.196.104.228:6666"` , then calico-node worked well even if network interface was restarted, and route was not been removed. 
May be that iptables of  k8s service is affected by restarting network interface.

**Comment by fasaxc** [comment]
@mdianjun That should work as a workaround.  

- If the NAT rule is in place, packets will be forwarded to the etcd pod.
- If the NAT rule is missing/flaps then the host will get the packet and respond with a RST.

**Comment by caseydavenport** [comment]
I think the root of the problem here was that Felix didn't properly re-establish its connection to etcd, as show in https://github.com/projectcalico/calico/issues/1632

This should be fixed in Calico v3.0.2 - please post back if you still see this in that version. Thanks!
